### ISSUE TITLE
VCF parsing fails for QUAL fields

### ISSUE THREAD

**Issue by nrizzo** [tsnorri/vcf2multialign#5]
Hey Tuukka!

I'm testing vcf2multialign (1.1 6e4c2d5) with the [1000 Genome Project data recalled on the T2T reference](https://github.com/marbl/CHM13). The VCF parsing stops with an error when reading a VCF entry containing a QUAL value written in scientific notation (e.g. **1.00735e+06**), and works otherwise when the values are integers.

```
$ vcf2multialign -r chr22.fasta -c chr22 -a test.vcf -f variant.graph
Assertions have been enabled.
[11:24:48] Reading the first reference sequence from the input FASTA… Done. Reference length is 22334.
[11:24:48] Building the variant graph…
Unexpected character 'e' (101) at 81:27, state 49.
** Last 128 charcters from the buffer:
e+06	PASS	AC=0;AF=0.262;AN=2;BaseQRankSum=0;DP=148660;ExcessHet=160;FS=2.714;InbreedingCoeff=-0.3113;MLEAC=1676;MLEAF=0.262;MQ=2
ERROR: Caught an exception: Unexpected character
```

I'm attaching a minimal example [vcf_parsing_issue.zip](https://github.com/tsnorri/vcf2multialign/files/15197779/vcf_parsing_issue.zip), obtained from the 1KGP+T2T dataset. Temporary fix is removing these quality values (vcf2multialign does not use QUAL, right?), I have approximated that with `sed 's/e+06//g'` but I'm not sure on how to do it nicely with bcftools.

Best,
Nicola

**Comment by tsnorri** [comment]
Hi! It's definitely a bug since the [QUAL field rules](https://github.com/tsnorri/libbio/blob/d01803f719201700f164a2bc83388bd3b10ed7dc/src/vcf_reader_parser.rl#L435-L438) in the VCF parser do not handle all the cases given in the VCF specification. Currently the most precise encoding rules are listed in Section 1.3 of the [VCF 4.3 specification](https://samtools.github.io/hts-specs/VCFv4.3.pdf). I’ll try to fix this as soon as I can.

**Comment by nrizzo** [comment]
Oops, I was using a vcf file containing **unphased** variations 😳 
Using a [correct dataset](https://github.com/JosephLalli/phasing_T2T) (phased with [SHAPEIT5](https://github.com/odelaneau/shapeit5)) another parsing issue comes up though:
```
$ vcf2multialign -r chr22.fasta -c chr22 -a test.vcf -f variant.graph
Assertions have been enabled.
[18:58:06] Reading the first reference sequence from the input FASTA… Done. Reference length is 463000.
[18:58:06] Building the variant graph…
WARNING: The definition of the info field “AC” in the VCF headers differs from that in the predefined fields (Integer, 1 vs. Integer, A).
ERROR: Caught an exception: More values in INFO field than declared
```
and once more, I approximated a fix with `sed 's/INFO=<ID=AC,Number=1/INFO=<ID=AC,Number=A/g'`. Attached is another minimal example  [vcf_parsing_issue_2.zip](https://github.com/tsnorri/vcf2multialign/files/15487221/vcf_parsing_issue_2.zip).

Thanks,
Nicola

**Comment by tsnorri** [comment]
I think the INFO field issue is unrelated. [VCF 4.3](https://samtools.github.io/hts-specs/VCFv4.3.pdf) (§ 1.4.2) requires that the INFO field header specifies the correct number of values that the field may contain in a given record. [VCF 4.2](https://samtools.github.io/hts-specs/VCFv4.2.pdf) is a bit more lenient about this (the word “should” is used in the corresponding section 1.2.2 instead of “must”) but vcf2multialign’s parser attempts to optimise memory use anyway and thus requires the field to have been declared properly.

**Comment by tsnorri** [comment]
QUAL field parsing is fixed in version 1.1.1. For the other problem, I think the correct approach is to fix the VCF header like you already did.
